File: 2015-16/contest3/prog/junior/abc.cpp
#include <bits/stdc++.h>
using namespace std;

bool match_given(char left[3][3], char right[3][3]) {
	for(int r = 0; r < 3; ++r)
		for(int c = 0; c < 3; ++c)
			if(left[r][c] && right[r][c] && left[r][c] != right[r][c])
				return false;
	return true;
}

bool just_once(char g[3][3]) {
	for(int r = 0; r < 3; ++r)
		for(int c = 0; c < 3; ++c) {
			for(int i = 0; i < 3; ++i) // check row r
				if(i != c && g[r][i] == g[r][c])
					return false;
			for(int i = 0; i < 3; ++i) // check col c
				if(i != r && g[i][c] == g[r][c])
					return false;
		}
	return true;
}

void print(char g[3][3]) {
	for(int r = 0; r < 3; ++r)
		for(int c = 0; c < 3; ++c)
			cout << g[r][c];
	cout << endl;
}

bool solve(char grid[3][3]) {
	vector<char> abc{'A', 'B', 'C'};
	for(char c1 : abc)
	for(char c2 : abc)
	for(char c3 : abc)
	for(char c4 : abc)
	for(char c5 : abc)
	for(char c6 : abc)
	for(char c7 : abc)
	for(char c8 : abc)
	for(char c9 : abc) {
		char g[3][3] = { {c1, c2, c3}, {c4, c5, c6}, {c7, c8, c9} };
		if(!match_given(g, grid))
			continue;
		if(!just_once(g))
			continue;
		// Found one solution
		for(int r = 0; r < 3; ++r)
			for(int c = 0; c < 3; ++c)
				grid[r][c] = g[r][c];
		return true;
	}
	return false;
}

int main() {
	for(int n = 0; n < 5; ++n) {
		char grid[3][3] = { {}, {}, {} };
		int N;
		cin >> N;
		for(int i = 0; i < N; ++i) {
			int p; char c;
			cin >> p >> c;
			--p;
			grid[p/3][p%3] = c;
		}
		if(solve(grid))
			print(grid);
		else
			cout << "Solution not found!" << endl;
	}
	return 0;
}

File: 2015-16/contest3/prog/junior/abc.py
import os
import sys

def match_given(left, right):
	for(int r = 0; r < 3; ++r)
		for(int c = 0; c < 3; ++c)
			if left[r][c] and right[r][c] and left[r][c] != right[r][c]:
				return False
	return True

def just_once(g):
	for r in range(3):
		for c in range(3):
			for i in range(3):
				if i != c and g[r][i] == g[r][c]:
					return False
			for i in range(3):
				if i != r and g[i][c] == g[r][c]:
					return False
	return True

def print(g):
	for r in range(3):
		for c in range(3):
			print(g[r][c], end='')
	print()
	return

def solve(grid):
	abc = ['A', 'B', 'C']
	for c1 in abc:
		for c2 in abc:
			for c3 in abc:
				for c4 in abc:
					for c5 in abc:
						for c6 in abc:
							for c7 in abc:
								for c8 in abc:
									for c9 in abc:
										g = [ [c1, c2, c3], [c4, c5, c6], [c7, c8, c9] ]
										if not match_given(g, grid):
											continue
										if not just_once(g):
											continue
										# Found one solution
										for r in range(3):
											for c in range(3):
												grid[r][c] = g[r][c]
										return True
	return False

if __name__=="__main__":
	for n in range(5):
		grid = [ [' ']*3 for _ in range(3)] # Create 3x3 with all spacers
		items = [ x for x in input().split()]
		N = int(items[0])
		for i in range(N):
			int p; char c;
			cin >> p >> c;
			--p;
			grid[p/3][p%3] = c;
		}
		if(solve(grid))
			print(grid);
		else
			cout << "Solution not found!" << endl;
